pyflakes/checker.py

```diff
@@ -518,6 +534,63 @@ def is_typing_overload_decorator(node):
     )
 
 
+def make_tokens(code):
+    # PY3: tokenize.tokenize requires readline of bytes
+    if not isinstance(code, bytes):
+        code = code.encode('UTF-8')
+    lines = iter(code.splitlines(True))
+    # next(lines, b'') is to prevent an error in pypy3
+    return tuple(tokenize_tokenize(lambda: next(lines, b'')))
+
+
+class _TypeableVisitor(ast.NodeVisitor):
+    """Collect the line number and nodes which are deemed typeable by
+    PEP 484
+
+    https://www.python.org/dev/peps/pep-0484/#type-comments
+    """
+    def __init__(self):
+        self.typeable_lines = []  # type: List[int]
+        self.typeable_nodes = {}  # type: Dict[int, ast.AST]
+
+    def _typeable(self, node):
+        # if there is more than one typeable thing on a line last one wins
+        self.typeable_lines.append(node.lineno)
+        self.typeable_nodes[node.lineno] = node
+
+        self.generic_visit(node)
+
+    visit_Assign = visit_For = visit_FunctionDef = visit_With = _typeable
+    visit_AsyncFor = visit_AsyncFunctionDef = visit_AsyncWith = _typeable
+
+
+def _collect_type_comments(tree, tokens):
+    visitor = _TypeableVisitor()
+    visitor.visit(tree)
+
+    type_comments = collections.defaultdict(list)
+    for tp, text, start, _, _ in tokens:
+        if (
+                tp != tokenize.COMMENT or  # skip non comments
+                not TYPE_COMMENT_RE.match(text) or  # skip non-type comments
```

Comment:
Do we want match or search? I wonder, can mypy/typed_ast handle something like:


```py
foo = 1   # noqa: E11  # type: ignore
```

Reply:
typed-ast is picky and requires the type comment to come first:

(using [astpretty](https://github.com/asottile/astpretty))

```
$ cat t.py 
x = 1  # noqa  # type: int
$ astpretty --typed-27 t.py 
Module(
    body=[
        Assign(
            lineno=1,
            col_offset=0,
            targets=[Name(lineno=1, col_offset=0, id='x', ctx=Store())],
            value=Num(lineno=1, col_offset=4, n=1),
            type_comment=None,
        ),
    ],
    type_ignores=[],
)
```

```
$ cat t2.py 
x = 1  # type: int  # noqa
$ astpretty --typed-27 t2.py
Module(
    body=[
        Assign(
            lineno=1,
            col_offset=0,
            targets=[Name(lineno=1, col_offset=0, id='x', ctx=Store())],
            value=Num(lineno=1, col_offset=4, n=1),
            type_comment='int  # noqa',
        ),
    ],
    type_ignores=[],
)
```

Reply:
Sorry, you should look a bit more closely at the second AST you pretty-printed, namely:

```
            type_comment='int  # noqa',
```

In other words, typed AST doesn't handle any case where some other linter's comment-based directives are present.

Reply:
it works, the inner comments get discarded since they're re-parsed by the ast:

```
$ cat t.py
from typing import List
x = []  # type: List[int]  # noqa
reveal_type(x)
$ mypy t.py
t.py:3: error: Revealed type is 'builtins.list[builtins.int]'
```